// @rcsb/rcsb-molstar/build/src/viewer/helpers/plddt-confidence/behavior.js
"use strict";
/**
 * Copyright (c) 2021 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Mandar Deshpande <mandar@ebi.ac.uk>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PLDDTConfidenceScore = void 0;
var tslib_1 = require("tslib");
var int_1 = require("molstar/lib/mol-data/int");
var prop_1 = require("./prop");
var color_1 = require("./color");
var structure_1 = require("molstar/lib/mol-model/structure");
var param_definition_1 = require("molstar/lib/mol-util/param-definition");
var behavior_1 = require("molstar/lib/mol-plugin/behavior/behavior");
exports.PLDDTConfidenceScore = behavior_1.PluginBehavior.create({
    name: 'plddt-confidence-prop',
    category: 'custom-props',
    display: {
        name: 'pLDDT Confidence Score',
        description: 'pLDDT Confidence Score.'
    },
    ctor: /** @class */ (function (_super) {
        (0, tslib_1.__extends)(class_1, _super);
        function class_1() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.provider = prop_1.PLDDTConfidenceProvider;
            _this.labelProvider = {
                label: function (loci) {
                    if (!_this.params.showTooltip)
                        return;
                    switch (loci.kind) {
                        case 'element-loci':
                            if (loci.elements.length === 0)
                                return;
                            var e = loci.elements[0];
                            var u = e.unit;
                            if (!u.model.customProperties.hasReference(prop_1.PLDDTConfidenceProvider.descriptor))
                                return;
                            var se = structure_1.StructureElement.Location.create(loci.structure, u, u.elements[int_1.OrderedSet.getAt(e.indices, 0)]);
                            var confidenceScore = prop_1.PLDDTConfidence.getConfidenceScore(se);
                            return confidenceScore ? "Confidence score: ".concat(confidenceScore[0], " <small>( ").concat(confidenceScore[1], " )</small>") : "No confidence score";
                        default: return;
                    }
                }
            };
            return _this;
        }
        class_1.prototype.register = function () {
            this.ctx.customModelProperties.register(this.provider, this.params.autoAttach);
            this.ctx.managers.lociLabels.addProvider(this.labelProvider);
            this.ctx.representation.structure.themes.colorThemeRegistry.add(color_1.PLDDTConfidenceColorThemeProvider);
        };
        class_1.prototype.update = function (p) {
            var updated = this.params.autoAttach !== p.autoAttach;
            this.params.autoAttach = p.autoAttach;
            this.params.showTooltip = p.showTooltip;
            this.ctx.customModelProperties.setDefaultAutoAttach(this.provider.descriptor.name, this.params.autoAttach);
            return updated;
        };
        class_1.prototype.unregister = function () {
            this.ctx.customModelProperties.unregister(prop_1.PLDDTConfidenceProvider.descriptor.name);
            this.ctx.managers.lociLabels.removeProvider(this.labelProvider);
            this.ctx.representation.structure.themes.colorThemeRegistry.remove(color_1.PLDDTConfidenceColorThemeProvider);
        };
        return class_1;
    }(behavior_1.PluginBehavior.Handler)),
    params: function () { return ({
        autoAttach: param_definition_1.ParamDefinition.Boolean(false),
        showTooltip: param_definition_1.ParamDefinition.Boolean(true)
    }); }
});
