// @rcsb/rcsb-molstar/build/src/viewer/ui/export.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExportControls = void 0;
var tslib_1 = require("tslib");
var jsx_runtime_1 = require("react/jsx-runtime");
var base_1 = require("molstar/lib/mol-plugin-ui/base");
var common_1 = require("molstar/lib/mol-plugin-ui/controls/common");
var icons_1 = require("molstar/lib/mol-plugin-ui/controls/icons");
var export_1 = require("../helpers/export");
var ExportControls = /** @class */ (function (_super) {
    tslib_1.__extends(ExportControls, _super);
    function ExportControls() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ExportControls.prototype.defaultState = function () {
        return {
            header: 'Export',
            isCollapsed: true,
            brand: { accent: 'gray', svg: ExportOutlinedSvg }
        };
    };
    ExportControls.prototype.componentDidMount = function () {
        var _this = this;
        this.subscribe(this.plugin.managers.structure.hierarchy.behaviors.selection, function (sel) {
            _this.setState({ isHidden: sel.structures.length === 0 });
        });
    };
    ExportControls.prototype.renderControls = function () {
        return (0, jsx_runtime_1.jsx)("div", tslib_1.__assign({ className: 'msp-control-offset', style: { paddingTop: '1px' } }, { children: (0, jsx_runtime_1.jsx)(CoordinatesExportControls, {}) }));
    };
    return ExportControls;
}(base_1.CollapsableControls));
exports.ExportControls = ExportControls;
var CoordinatesExportControls = /** @class */ (function (_super) {
    tslib_1.__extends(CoordinatesExportControls, _super);
    function CoordinatesExportControls() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.download = function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            var content;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        content = (0, export_1.encodeStructureData)(this.plugin);
                        return [4 /*yield*/, (0, export_1.downloadAsZipFile)(this.plugin, content)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        }); };
        return _this;
    }
    CoordinatesExportControls.prototype.render = function () {
        return (0, jsx_runtime_1.jsx)(jsx_runtime_1.Fragment, { children: (0, jsx_runtime_1.jsx)("div", tslib_1.__assign({ className: 'msp-flex-row' }, { children: (0, jsx_runtime_1.jsx)(common_1.Button, tslib_1.__assign({ icon: icons_1.GetAppSvg, onClick: this.download, title: 'Save structures as mmCIF files' }, { children: "Structures" })) })) });
    };
    return CoordinatesExportControls;
}(base_1.PluginUIComponent));
function ExportOutlinedSvg() { return _ExportOutlined; }
var _ExportOutlined = (0, jsx_runtime_1.jsx)("svg", tslib_1.__assign({ width: '24px', height: '24px', viewBox: '0 0 24 24', strokeWidth: '0.1px' }, { children: (0, jsx_runtime_1.jsx)("path", { d: "M19 12v7H5v-7H3v9h18v-9h-2zm-6 .67l2.59-2.58L17 11.5l-5 5-5-5 1.41-1.41L11 12.67V3h2v9.67z" }) }));
